### PR TITLE
Support IEEE-754 for fmin/fmax/dmin/dmax nodes

### PR DESCRIPTION
**Pull request by matthewhall2**
- Enables inlining of fmin/fmax/dmin/dmax nodes
- Implements IEEE-754 standard for the evaluators:
  - if the first arg is a NaN, returns the corresponding quiet NaN, same for if only the second arg is a NaN
-  Refactors `xmaxxminhelper` to infer correct compare/branch/move instructions based on datatype; helper does not change NaNs, so can be used in other min/max evaluators which can extend the behaviour on NaNs as they see fit
- Uses vector operations when available, defaults to branching if not
- Supports inlining for https://github.com/eclipse-openj9/openj9/pull/20185

Depends on OMR Simplifier change https://github.com/eclipse/omr/pull/7471 (needed for https://github.com/eclipse-openj9/openj9/pull/20185 to pass)

https://hyc-runtimes-jenkins.swg-devops.com/job/jvm.29.personal/34283/

### CODE REVIEW

**Comment by matthewhall2** [comment]
can you review please @r30shah ?

File: compiler/z/codegen/OMRTreeEvaluator.hpp
Comment (r30shah) [inline compiler/z/codegen/OMRTreeEvaluator.hpp:256]: Either here or where the function is implemented, we should add a doxygen styled comments for function description(Take a look at functions in z/OMRTreeEvaluator.hpp [1])

[1]. https://github.com/eclipse/omr/blob/13d47ea4def377422793a23813b6b25c76bbd56d/compiler/z/codegen/OMRTreeEvaluator.hpp#L924-L951

File: compiler/z/codegen/ControlFlowEvaluator.cpp
Comment (r30shah) [inline compiler/z/codegen/ControlFlowEvaluator.cpp:678]: We should add a brief explanation for these Load Test instructions.

File: compiler/z/codegen/ControlFlowEvaluator.cpp
Comment (r30shah) [inline compiler/z/codegen/ControlFlowEvaluator.cpp:328]: I think the original name `xmaxminHelper` should be ok (You missed the caps in Helper when refactoring).

**Review by r30shah — CHANGES_REQUESTED** [review]

File: compiler/z/codegen/ControlFlowEvaluator.cpp
Comment (r30shah) [inline compiler/z/codegen/ControlFlowEvaluator.cpp:363]:
```suggestion
      /**
       * Check for NaN operands for float and double
       * Support float and double +0/-0 comparisons adhering to IEEE 754 standard
       * Checking if operands are equal, then branching to equalRegion, otherwise
       * fall through for NaN case handling
       */
```

File: compiler/z/codegen/ControlFlowEvaluator.cpp
Comment (r30shah) [inline compiler/z/codegen/ControlFlowEvaluator.cpp:421]: We should combine this two asserts.

File: compiler/z/codegen/OMRTreeEvaluator.hpp
Comment (r30shah) [inline compiler/z/codegen/OMRTreeEvaluator.hpp:259]: I think we should mention that it generates Vector Instructions when available for max/min operations.

**Review by r30shah — CHANGES_REQUESTED** [review]
Last nitpicks

**Comment by r30shah** [comment]
Re, three commits you have, I think https://github.com/eclipse/omr/pull/7464/commits/fe772f76faf1043a0ca2c8599f0c23176b2df279 makes sense.

For https://github.com/eclipse/omr/pull/7464/commits/a95eed52c584d5ae2f24c812ebcd60c04af95f5f - Can we reword the description. To something that emphasize that the change does not alter the OMR implementation, but adds changes to helper function for max min to support Java specification for such operations.

File: compiler/z/codegen/ControlFlowEvaluator.cpp
Comment (r30shah) [inline compiler/z/codegen/ControlFlowEvaluator.cpp:39]: This should be moved down with other includes from `z/codegen`

**Review by r30shah — APPROVED** [review]
Approving the change, one last comment about the Include, otherwise looks good to me.

**Review by r30shah — APPROVED** [review]
I think this change which Matthew inherited from Sarwat are ok as well, but @hzongaro  can you share what do you think ? 


@matthewhall2  can you share the other related OpenJ9 PR as also confirm if this can be delivered independantly? 
Also codegen changes are OK, so can we request you to review this PR as well?

File: compiler/z/codegen/ControlFlowEvaluator.cpp
Comment (matthewhall2) [reply compiler/z/codegen/ControlFlowEvaluator.cpp]: done

**Comment by matthewhall2** [comment]
@r30shah @hzongaro here's the [openj9 change](https://github.com/eclipse-openj9/openj9/pull/20185). 
It can be delivered independently (it will compile and run), but OpenJ9 will follow the IEEE spec and not the Java spec until the openj9 PR is merged

**Comment by r30shah** [comment]
> This PR has to be merged first. The only potential issue is that OpenJ9 will follow the IEEE spec and not the Java spec until the openj9 PR is merged

Looking at your openJ9 change, is this statement true? Without OpenJ9 change , Java should not be inlining max and min for floating point types right >?

**Comment by matthewhall2** [comment]
> Looking at your openJ9 change, is this statement true? Without OpenJ9 change , Java should not be inlining max and min for floating point types right >?

ah yes, they won't be inlined without the openj9 change. But the opnej9 PR won't compile without this PR, so this one still needs to be merged first.

**Comment by matthewhall2** [comment]
@hzongaro can you review when you get the chance?

**Review by hzongaro — COMMENTED** [review]
I think the changes look good overall.  Just a few minor comments.

May I also ask you to update the commit comments to indicate that these changes are for inline code for z/Architecture only?

File: compiler/z/codegen/ControlFlowEvaluator.cpp
Comment (hzongaro) [inline compiler/z/codegen/ControlFlowEvaluator.cpp:364]: You should change this back to either of the following styles of comment blocks:

```
  //
  //
  //
```

or

```
/*
 *
 */
```

Sorry I didn't comment on this in my earlier review, but I had to double-check Doxygen's behaviour.  At least some versions of Doxygen will pick up `/** */` style comments as Doxygen comments even in the middle of a method.

File: compiler/z/codegen/ControlFlowEvaluator.cpp
Comment (hzongaro) [inline compiler/z/codegen/ControlFlowEvaluator.cpp:422]: Now that I've taken a second look at this assertion, should it be `node->getNumChildren() == 2 && node->getOpCode().isFloatingPoint()` instead?  The call to `node->getSecondChild()` below would cause problems if `node->getNumChildren()` returned one.

**Review by hzongaro — COMMENTED** [review]
Looks good.  Just one further minor comment.

File: compiler/z/codegen/ControlFlowEvaluator.cpp
Comment (matthewhall2) [reply compiler/z/codegen/ControlFlowEvaluator.cpp]: done. I also changed the branching version of the helper to use an additional branch instruction in the floating point path. This allows it to skip over some of the instructions that deal with NaNs and improves performance by about 4% (in addition to the ~17% improvement that was already in this PR). 
please review @r30shah @hzongaro

Comment (r30shah) [reply compiler/z/codegen/ControlFlowEvaluator.cpp]: Hmm, Addiing an additional branch may interfere with the prediction unit. I assume you are seeing 3% in the bumblebench, is that consistent ?

Comment (matthewhall2) [reply compiler/z/codegen/ControlFlowEvaluator.cpp]: yup it's consistent. I ran the bumblebench benchmark with multiple different seeds for the RNG, multiple times per seed for each version. The one with the extra instruction consistently gets a score of 96M, and with no extra a score of 93M. Both have an uncertainty of 0.0% or 0.1% most of the time, and never above 1%

Comment (matthewhall2) [reply compiler/z/codegen/ControlFlowEvaluator.cpp]: > Hmm, Addiing an additional branch may interfere with the prediction unit. I assume you are seeing 3% in the bumblebench, is that consistent ? 
> 
> Also if I am reading correctly, we would have branch like following, 
> ```
> BRC  returnFirstArgCond, done
> BRC  returnSecondArgCond, swapValues
> BRC COND_8, equalOperands
> ...
> ```
> 
> 

Extra branch instruction has been removed as discussed

**Review by r30shah — APPROVED** [review]
Thanks @matthewhall2  for the changes, it looks good to me, will wait for @hzongaro  to approve and merge this before I go on and look at the J9 changes.

**Comment by r30shah** [comment]
jenkins build zos,zlinux

**Review by hzongaro — APPROVED** [review]
Looks good.  Thanks for addressing the comments!

**Pull request merged**
